lib/execution_segment.go

```diff
@@ -394,13 +396,275 @@ func (ess ExecutionSegmentSequence) String() string {
+		numerator     int64
+		originalIndex int
+	}
+	lcd int64
+}
+
+func newWrapper(ess ExecutionSegmentSequence) sortInterfaceWrapper {
+	var result = sortInterfaceWrapper{
+		slice: make([]struct {
+			numerator     int64
+			originalIndex int
+		}, len(ess)),
+		lcd: ess.lcd(),
+	}
+
+	for i := range ess {
+		result.slice[i].numerator = ess[i].length.Num().Int64() * (result.lcd / ess[i].length.Denom().Int64())
+		result.slice[i].originalIndex = i
+	}
+
+	sort.SliceStable(result.slice, func(i, j int) bool {
+		return result.slice[i].numerator > result.slice[j].numerator
+	})
+	return result
+}
+
+// Imagine you have a number of rational numbers which all add up to 1 (or less) and call them
+// segments.
+// If you want each to get proportional amount of anything you need to give them their numerator
+// count of elements for each denominator amount from the original elements. So for 1/3 you give 1
+// element for each 3 elements. For 3/5 - 3 elements for each 5.
+// If you have for example a sequence of with element with length 3/5 and 1/3 in order to know how
+// to distribute it accurately you need to get the LCD(lowest common denominitor) in this case
+// between 3 and 5 this is 15 and then to transform the numbers to have the same, LCD equal,
+// denominator. So 3/5 becomes 9/15 and 1/3 becomes 5/15. So now for each 15 elements 9 need to go
+// to the 3/5, and 5 need to go to 1/3.
+//
+// We use the below algorithm to split elements between ExecutionSegments by using their length as
+// the rational number. As we would like to get non sequential elements we try to get the maximum
+// distance between them. That is the number of elements divided by the number of elements for any
+// given segment, which concidently is the length of the segment reversed.
+// The algorithm below does the following:
+// 1. Goes through the elements from 0 to the lcd-1
+// 2. For each of element goes through the segments and looks if the amount of already taken
+// elements by the given segment multiplied by that segment length inverted is equal to or less to
+// the current element index. if it is give that element to that segment if not continue with the
+// next element.
 //
-// TODO: add a more detailed algorithm description
-func (ess ExecutionSegmentSequence) GetStripedOffsets(segment *ExecutionSegment) (int, []int, error) {
-	start := 0
-	offsets := []int{}
-	// TODO: basically https://docs.google.com/spreadsheets/d/1V_ivN2xuaMJIgOf1HkpOw1ex8QOhxp960itGGiRrNzo/edit
-	return start, offsets, fmt.Errorf("not implemented")
+// The code below specifically avoids using big.Rat which complicates the code somewhat.
+// As additional note the sorting of the segments from biggest to smallest helps with the fact that
+// the biggest elements will need to take the most elements and for them it will be the hardest to
+// not get sequential elements.
+func (e sortInterfaceWrapper) stripingAlgorithm(saveIndex func(iteration int64, index int, numerator int64) bool) {
```

Comment:
:+1: for dropping `Algorithm` and `stripe` sounds even better.

This is for 2 reasons:
1. This is artefact of how it was previously and during the rewrites. As we are running out of time and this is only part of the work needed, and I needed feedback ... in more general sense and there is also
2. While this currently isn't used as I calculate all offsets once and then cache it ... maybe we will want to revert to using the other variant - this will depend both on how much we think that caching actually speedups stuff and what we need in the other executors ... which are both connected. 

So I would prefer if we add a TODO to refactor this once we have fixed all the functional issues and probably have gain some experience about what is needed.